### ISSUE TITLE
Confidence intervals with `predictions()` and `nnet::multinom`

### ISSUE THREAD

**Issue by tomwenseleers** [vincentarelbundock/marginaleffects#404]
I would like to obtain lower and upper 95% confidence intervals of `nnet:multinom` models.
If I have a `nnet:multinom` model `nnet_fit`
and execute `preds = predictions(nnet_fit, type="probs")` I only get predictions and standard errors though.

And `preds = predictions(nnet_fit, type="link", transform_post = insight::link_inverse(nnet_fit))` gives the error
"Error: The `type` argument for models of class `multinom` must be an element of: probs".

Any thought on how to obtain these?

**Comment by vincentarelbundock** [comment]
Thanks for the report.

The main problem is that `marginaleffects::predictions()` relies on the `predict()` methods supplied by the modeling packages to compute predicted values. In this case, the `predict()` method supplied by `nnet` does not support predictions on the link scale:

``` r
library(nnet)
library(marginaleffects)

mod <- multinom(factor(gear) ~ mpg + hp, data = mtcars, trace = FALSE)

# works
predict(mod, type = "probs") |> head()
#>                            3          4           5
#> Mazda RX4         0.29615334 0.67938691 0.024459743
#> Mazda RX4 Wag     0.29615334 0.67938691 0.024459743
#> Datsun 710        0.08403718 0.89906868 0.016894140
#> Hornet 4 Drive    0.25519341 0.71283439 0.031972199
#> Hornet Sportabout 0.72122841 0.04380171 0.234969877
#> Valiant           0.59881931 0.39937632 0.001804366

# breaks
predict(mod, type = "link") |> head()
#> Error in match.arg(type): 'arg' should be one of "class", "probs"
```

Since we can’t easily get predictions on the link scale, we can’t do what we usually do: build confidence intervals by back-transformation.

One alternative would be to use the internal (unexported) function `get_ci()` to simply multiply the standard errors by a critical Z value. However, the resulting intervals will not have the same nice properties as the back-transformed ones. For example, you can see negative bounds here:

``` r
p <- predictions(mod, type = "probs")
marginaleffects:::get_ci(p, conf_level = .9, estimate = "predicted") |> head()
#>   rowid  type group  predicted  std.error gear  mpg  hp statistic      p.value    conf.low conf.high
#> 1     1 probs     3 0.29615334 0.15269449    4 21.0 110  1.939516 5.243859e-02  0.04499325 0.5473134
#> 2     2 probs     3 0.29615334 0.15269449    4 21.0 110  1.939516 5.243859e-02  0.04499325 0.5473134
#> 3     3 probs     3 0.08403718 0.08257637    4 22.8  93  1.017690 3.088251e-01 -0.05178887 0.2198632
#> 4     4 probs     3 0.25519341 0.14886321    3 21.4 110  1.714281 8.647708e-02  0.01033522 0.5000516
#> 5     5 probs     3 0.72122841 0.16413979    3 18.7 175  4.393989 1.112895e-05  0.45124248 0.9912143
#> 6     6 probs     3 0.59881931 0.31123458    3 18.1 105  1.924013 5.435298e-02  0.08688400 1.1107546
```

Unfortunately, I don’t think it’s realistic for me to implement `predict()` methods for all the 60+ model types supported, so we have to rely to some extent on the modeling packages being “feature-complete”…

**Comment by tomwenseleers** [comment]
Thanks! Ha that is a shame... Do you think it might be worth allowing passing a custom predict (+vcov) method function in case the original package author is not willing to update the code in the package? Or can one redefine such methods, overwriting the original method defined in the original package (also asked that question here [https://stackoverflow.com/questions/73010776/redefining-rs-nnetmultinom-predict-multinom-predict-method-to-support-type-l]? It might be a good idea to have a proper check of all the models with link scales supposedly supported by `marginaleffects `to see if they are all able to correctly work on that scale, as I really don't think it makes sense (e.g. report standard errors) when calculated on the response scale...

Below I put a reproducible example of what I would like to achieve (in terms of calculating marginal means or predictions with confidence intervals as well as calculating contrasts of marginal trends, which I wasn't sure if that is currently supported in principle), with the output that I get with `emmeans `and `emtrends`, and using multinomial models fit with either `nnet::multinom, mclogit::mblogit` or `VGAM::vglm`. I haven't tried with `mlogit` (or the faster version `mnlogit`), as that seems to require a slightly different data structure, so couldn't readily figure that one out. Not sure if there is any other options that eventually could be supported by `marginaleffects `- `mgcv::gam, glmnet::glmnet, brms::brm, bamlss::bamlss or bayesCL::mlr` perhaps - but not sure? 

```
# REPRODUCIBLE EXAMPLES multinomial regression with nnet:multinom, mclogit::mblogit, VGAM::vglm + emmeans & marginaleffects tests

# data=SARS-CoV2 coronavirus variants (variant) through time (collection_date_num)
# in India, count=actual count (nr of sequenced genomes)

dat = read.csv("https://www.dropbox.com/s/u27cn44p5srievq/dat.csv?dl=1")
dat$variant = factor(dat$variant)

# 1. multinom::net multinomial fit ####
library(nnet)
library(splines)
set.seed(1)
fit_nnet = nnet::multinom(variant ~ ns(collection_date_num, df=2), 
                          weights=count, data=dat)
summary(fit_nnet)

# emmeans & emtrends using emmeans package: works, but slow
# emmeans = current proportion of different variants
library(emmeans)
multinom_emmeans = emmeans(fit_nnet, ~ variant,  
                       mode = "prob",
                       at=list(collection_date_num = 
                                 max(data_agbyweek1$collection_date_num)))
multinom_emmeans
# variant               prob       SE df lower.CL upper.CL
# Alpha             0.00e+00 0.00e+00 33 0.00e+00 0.00e+00
# Beta              0.00e+00 0.00e+00 33 0.00e+00 0.00e+00
# Delta             7.73e-06 1.17e-06 33 5.34e-06 1.01e-05
# Omicron (BA.1)    1.82e-04 6.42e-05 33 5.14e-05 3.13e-04
# Omicron (BA.2)    1.76e-01 7.45e-03 33 1.61e-01 1.91e-01
# Omicron (BA.2.74) 9.03e-02 7.98e-03 33 7.41e-02 1.07e-01
# Omicron (BA.2.75) 1.68e-01 1.90e-02 33 1.30e-01 2.07e-01
# Omicron (BA.2.76) 2.89e-01 1.35e-02 33 2.62e-01 3.16e-01
# Omicron (BA.3)    1.34e-02 2.10e-03 33 9.10e-03 1.76e-02
# Omicron (BA.4)    1.67e-02 2.47e-03 33 1.17e-02 2.17e-02
# Omicron (BA.5)    2.03e-01 1.08e-02 33 1.81e-01 2.25e-01
# Other             4.23e-02 3.15e-03 33 3.59e-02 4.87e-02
# 
# Confidence level used: 0.95 

# pairwise contrasts in marginal trends on link scale
# = current differences in growth rate among variants
multinom_emtrends = confint(pairs(emtrends(fit_nnet, ~ variant,  
                                       var = "collection_date_num",  
                                       mode = "latent",
                                       at = list(collection_date_num = 
                                                 max(data_agbyweek1$collection_date_num))),
                              reverse=TRUE))
multinom_emtrends

# equivalent using margineffects package 
# does not work as there is no predict.multinom method with type="link"
library(marginaleffects)
multinom_preds_marginaleffects = predictions(fit_nnet,
                                         newdata = datagrid(collection_date_num = 
                                                              max(data_agbyweek1$collection_date_num)),
                                         type="link", # not supported by predict.multinom
                                         transform_post = insight::link_inverse(fit_nnet))

# Error: The `type` argument for models of class `multinom` must be an element of: probs
multinom_preds_marginaleffects 

# NOTE: would it be possible perhaps to redefine the predict.multinom function to use the
# predict.mblogit method instead, which might work after small modification,
# and which does support type="link":
# https://stackoverflow.com/questions/73010776/redefining-rs-nnetmultinom-predict-multinom-predict-method-to-support-type-l

# marginal effects
# same problem for marginal effects
multinom_marginaleffects = marginaleffects(fit_nnet, newdata = datagrid(collection_date_num = 
                                             max(data_agbyweek1$collection_date_num)),
                variables="collection_date_num",
                # type="probs"
                type="link" # not supported by predict.multinom
                )
# Error: The `type` argument for models of class `multinom` must be an element of: probs

# NOTE: I was not sure if pairwise contrasts in marginal effects 
# as in emtrends example above are supported by marginaleffects & of syntax,
# comparisons(multinom_marginaleffects) e.g. does not work


# 2. mclogit::mblogit multinomial fit ####
# this one does have a predict.mblogit method with type="link",
# https://github.com/melff/mclogit/blob/master/pkg/R/mblogit.R
library(mclogit)
fit_mblogit = mblogit(variant ~ ns(collection_date_num, df=2),
                      weight=count,
                      data=dat,
                      from.table=TRUE, dispersion=FALSE,
                      control=mclogit.control(maxit=25))
summary(fit_mblogit)
# PS increasing maxit to >30 results in 
# Error in solve.default(Information) : 
#  system is computationally singular: reciprocal condition number = 1.66419e-16
# maybe due to complete separation / Hauck-Donner effect or bug in mblogit??
# Right now SEs are very large, probably bug in mblogit or due to H-D effect

# emmeans & emtrends using emmeans package: works, but slow
# emmeans = current proportion of different variants
mblogit_emmeans = emmeans(fit_mblogit, ~ variant,  
                           mode = "prob",
                           at=list(collection_date_num = 
                                     max(data_agbyweek1$collection_date_num)))
mblogit_emmeans
# variant             prob       SE  df asymp.LCL asymp.UCL
# Alpha             0.0000 9.00e-08 Inf -1.80e-07  2.00e-07
# Beta              0.0000 1.00e-08 Inf -2.00e-08  0.00e+00
# Delta             0.0000 1.00e-08 Inf -2.00e-08  0.00e+00
# Omicron (BA.1)    0.0000 1.00e-08 Inf -2.00e-08  0.00e+00
# Omicron (BA.2)    0.3653 3.73e-02 Inf  2.92e-01  4.38e-01
# Omicron (BA.2.74) 0.0299 1.32e-02 Inf  4.09e-03  5.58e-02
# Omicron (BA.2.75) 0.1916 3.05e-02 Inf  1.32e-01  2.51e-01
# Omicron (BA.2.76) 0.3473 3.68e-02 Inf  2.75e-01  4.20e-01
# Omicron (BA.3)    0.0000 1.00e-08 Inf -2.00e-08  0.00e+00
# Omicron (BA.4)    0.0120 8.42e-03 Inf -4.52e-03  2.85e-02
# Omicron (BA.5)    0.0539 1.75e-02 Inf  1.96e-02  8.81e-02
# Other             0.0000 1.00e-08 Inf -2.00e-08  0.00e+00
# 
# Confidence level used: 0.95
# NOTE: does not quite match emmeans nnet::multinom output
# above, but maybe because I couldn't run mblogit until convergence

# pairwise contrasts in marginal trends on link scale
# = current differences in growth rate among variants
# confidence intervals are here too wide, probably bug in mblogit or H-D effect
mblogit_emtrends = confint(pairs(emtrends(fit_mblogit, ~ variant,  
                                           var = "collection_date_num",  
                                           mode = "latent",
                                           at = list(collection_date_num = 
                                                       max(data_agbyweek1$collection_date_num))),
                                  reverse=TRUE))
mblogit_emtrends


# equivalent using margineffects package 
mblogit_preds_marginaleffects = predictions(fit_mblogit,
                                             newdata = datagrid(collection_date_num = 
                                                                  max(data_agbyweek1$collection_date_num)),
                                             type="link", # supposedly supported by predict.mblogit
                                             transform_post = insight::link_inverse(fit_nnet))
                                             #transform_post = insight::link_inverse(fit_mblogit)) # function (eta) eta is not correct, inverse link for multinomial should be softmax function
# returns Error in transform_post(draws) : 
# REAL() can only be applied to a 'numeric', not a 'NULL'
# so something going wrong with transform_post
# leaving that out gives predictions on link scale (I think)
# conf intervals too wide though - prob bug in mblogit or H-D effect
mblogit_preds_marginaleffects


# 3. VGAM::vglm multinomial fit ####
# this one supposedly does have a predict.vglm method with type="link",
# https://github.com/cran/VGAM/blob/master/R/predict.vglm.q
library(VGAM)
fit_vglm = vglm(variant ~ ns(collection_date_num, df=2),
                      weight=count+1E-5, # I added 1E-5 to prevent an error
                      family=multinomial,
                      data=dat)
summary(fit_vglm) # reports problems with Hauck-Donner effects
vglm_preds_marginaleffects = predictions(fit_vglm,
                                             newdata = datagrid(collection_date_num = 
                                                                  max(data_agbyweek1$collection_date_num)),
                                             type="response" # here = "probs" 
                                             # type="link", 
                                             # transform_post = insight::link_inverse(fit_vglm))
) 
# results in error Error in `*tmp*`[["coefficients"]] : this S4 class is not subsettable
vglm_preds_marginaleffects 
```

**Comment by vincentarelbundock** [comment]
One thing I could eventually do is allow a custom class name. Say you fit a `multinom` model, you could then do:

```r
class(model) <- c("marginaleffects_custom_class", class(model))
```

With class inheritance, this object should still be supported with all the current machinery. Internally, this means that the `marginaleffects:::get_predict()` methods would work and return an appropriate data frame. But then, a user could also define a new method called `get_predict.marginaleffects_custom_class()`, and that would take precedence over the existing ones, and would override it.

Would that work for you? Do you have a working `get_predict()` that works on the link scale?

**Comment by tomwenseleers** [comment]
I think that could work yes! I don't have a working `get_predict() `method that works on the link scale yet no - I sent Brian Ripley, the `nnet `package author, a message - but given that he's 70 and in bad health not sure if he would be able to add it... If I were to try myself I would probably try to base myself on the `predict.mblogit` method that does have `type="link"`,
[,](https://github.com/melff/mclogit/blob/master/pkg/R/mblogit.R) but I would have to make some minor changes, e.g. due to some slight differences in the `mblogit `& `multinom `object structures. The `mblogit `fit also didn't play well with `marginaleffects `btw - see above - there was an error there in `insight::link_inverse(fit_mblogit) `not giving the correct inverse link function, but I think without the `transform_post `argument it did give the right nrs but on a link scale (though in this particular case the `mblogit `fit itself didn't seem to match with the `nnet::multinom` one, maybe due some problem of complete separation).

**Comment by vincentarelbundock** [comment]
Sounds good.

I agree that the chances of this being implemented in `nnet` are close to zero.

Let me know if/when you make progress on this. I'm unlikely to implement this "just in case", but I'll do it if there's a concrete use-case.

Maybe you could report the `link_inverse()` problem on the `insight` repo: https://github.com/easystats/insight/issues

**Comment by tomwenseleers** [comment]
Thanks - I'll do, and I'll post a report on that bug in `link_inverse()` too! One more question re. the code I posted above: are pairwise contrasts in marginal trends, as in
```
mblogit_emtrends = confint(pairs(emtrends(fit_mblogit, ~ variant,  
                                           var = "collection_date_num",  
                                           mode = "latent",
                                           at = list(collection_date_num = 
                                                       max(data_agbyweek1$collection_date_num))),
                                  reverse=TRUE))
```
supported by `marginaleffects `(in principle, in this case if predictions on the link scale were working)?

**Comment by vincentarelbundock** [comment]
Yes, I think you want the `hypothesis` argument a
with the ˋ"pairwise"` option: https://vincentarelbundock.github.io/marginaleffects/articles/hypothesis.html#pairwise-contrasts-difference-in-differences

**Comment by vincentarelbundock** [comment]
I'm closing issues to "clean up" the repository in anticipation for a release. Don't hesitate to open an issue if/when you get to a point where accepting custom model classes would be useful in a real-life use-case.

Thanks again for the ideas and exchange!

**Comment by tomwenseleers** [comment]
So an easy solution to solve the `nnet:multinom` predict method lacking a `type="link`" `predict `method would be to create a wrapper method that calls the original method. For this we would also need to create a wrapper `multinom `function in the `marginaleffects `package, so that the correct package namespace is found. A sketch implementation I think would be (I also give the correct link & inverse link function for multinomial below - these were currently incorrect for both `multinom `& `mblogit`) :

```
multinom <- function(...)
{
    nnet::multinom(...)
}

# this is the link function for multinomial
# = generalized logit
inverse_softMax <- function(mu) {
  log_mu <- log(mu)
  return(sweep(log_mu, 1, STATS=rowMeans(log_mu), FUN="-")) # we let the log(odds) sum to zero - these predictions are referred to as type="latent" in the emmeans package
}

# this is the inverse link function for multinomial (here assuming eta is a matrix with different outcome levels as columns)
# = inverse generalized logit
softMax <- function(eta){
  exp_eta <- exp(eta)
  return(sweep(exp_eta, 1, STATS=rowSums(exp_eta), FUN="/"))
}

predict.multinom <- function(*, type=c("probs", "response", "latent", "link") # probs=response, latent=link (latent is how it is referred to in the emmeans package)
{
    type <- match.arg(type)
    if (type == "probs"|type == "response")
        return(nnet::predict.multinom(*, type="probs"))

    mu <- nnet::predict.multinom(*, type="probs")
    return(inverse_softMax(mu))

}
```

For `predict.mblogit` we also have the problem that `predict(fit_mblogit, type="link")` drops the first category (it normalizes the logits so that that level is zero, but doesn't explicitly return it). In the context of `marginaleffects `we would like all response levels to come out, and have the logits normalized so that they sum to one (as with `type="latent"` in the `emmeans` package). To change this behaviour we would also need 

```
mblogit <- function(...)
{
    mclogit::mblogit(...)
}

predict.mblogit <- function(*, type=c("probs", "response", "latent", "link") # probs=response, latent=link (latent is how it is referred to in the emmeans package)
{
    type <- match.arg(type)
    if (type == "probs"|type == "response")
        return(mclogit::predict.mblogit(*, type="response"))

    mu <- nnet::predict.mblogit(*, type="probs")
    return(inverse_softMax(mu))

}

```
The variance-covariance function returned by `mblogit `might also need checking - Russ Lenth alerted me that that one comes out in a non-standard order & in `emmeans `the `vcov `matrix of `mblogit `objects is returned in a permuted form, see [https://github.com/rvlenth/emmeans/blob/master/R/multinom-support.R](https://github.com/rvlenth/emmeans/blob/master/R/multinom-support.R):
```
    # we have to arrange the vcov elements in row-major order
    if(missing(vcov.))
        vcov. = vcov(object)
    perm = matrix(seq_along(as.numeric(object$coefmat)), 
                  ncol = ncol(object$coefmat))
    perm = as.numeric(t(perm))
    vcov. = vcov.[perm, perm]
```

Would it be possible to reopen this issue so that these things could be resolved, and multinomial support in `marginaleffects `would be fully functional?
